c++/src/capnp/compat/http-over-capnp.c++

```diff
@@ -688,6 +830,69 @@ public:
     });
   }
 
+  kj::Promise<void> connect(ConnectContext context) override {
+    auto params = context.getParams();
+    auto host = params.getHost();
+    auto headers = factory.headersToKj(params.getHeaders());
+    auto pipe = kj::newTwoWayPipe();
+
+    class EOFDetector final: public kj::AsyncOutputStream {
+    public:
+      EOFDetector(kj::Own<kj::AsyncIoStream> inner) : inner(kj::mv(inner)) {}
+      ~EOFDetector() {
+        inner->shutdownWrite();
+      }
+
+      kj::Maybe<kj::Promise<uint64_t>> tryPumpFrom(
+          kj::AsyncInputStream& input, uint64_t amount = kj::maxValue) override {
+        return inner->tryPumpFrom(input, amount);
+      }
+
+      kj::Promise<void> write(const void* buffer, size_t size) override {
+        return inner->write(buffer, size);
+      }
+
+      kj::Promise<void> write(kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) override {
+        return inner->write(pieces);
+      }
+
+      kj::Promise<void> whenWriteDisconnected() override {
+        return inner->whenWriteDisconnected();
+      }
+    private:
+      kj::Own<kj::AsyncIoStream> inner;
+    };
+
+    kj::Own<kj::AsyncOutputStream> stream = factory.streamFactory.capnpToKj(
+        context.getParams().getDown());
+    auto promises = kj::heapArrayBuilder<kj::Promise<void>>(2);
+    // We write to the `down` pipe.
+    promises.add(pipe.ends[1]->pumpTo(*stream).then([context](uint64_t) mutable {
+      auto endReq = context.getParams().getDown().endRequest();
+      return endReq.send().ignoreResult();
+    }).attach(kj::mv(stream)).eagerlyEvaluate(
+        [&httpProxyStream = *pipe.ends[1]](kj::Exception&&) { httpProxyStream.abortRead(); }));
+
+    PipelineBuilder<ConnectResults> pb;
+    auto eofWrapper = kj::heap<EOFDetector>(kj::mv(pipe.ends[1]));
+    auto up = factory.streamFactory.kjToCapnp(kj::mv(eofWrapper));
+    pb.setUp(kj::cp(up));
+
+    context.setPipeline(pb.build());
+    context.initResults(capnp::MessageSize { 4, 1 }).setUp(kj::mv(up));
+
+    class FinalHttpOverCapnpConnectResponseImpl final: public HttpOverCapnpConnectResponseImpl {
```

Comment:
I'm curious why this class is necessary?

Reply:
Good question and I am not sure myself. I followed how `request` above was implemented and it's using a similar class.

Reply:
Here you can just declare `HttpOverCapnpConnectResponseImpl` itself as `final`.

In the `request` case this wasn't possible, because `startRequest()` has its own subclass that works a bit differently.

Reply:
(And the reason things need to be declared `final` is because otherwise we get a compiler warning about calling a non-virtual destructor on a non-final polymorphic class...)